iceoryx_utils/include/iceoryx_utils/internal/cxx/storable_function.inl

```diff
@@ -0,0 +1,292 @@
+    // destroy is not needed for free functions
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename T, typename>
+storable_function<S, signature<ReturnType, Args...>>::storable_function(T& object,
+                                                                        ReturnType (T::*method)(Args...)) noexcept
+{
+    auto p = &object;
+    auto functor = [p, method](Args... args) -> ReturnType { return (*p.*method)(std::forward<Args>(args)...); };
+    storeFunctor(functor);
+}
+
+template <typename S, typename ReturnType, typename... Args>
+template <typename T, typename>
+storable_function<S, signature<ReturnType, Args...>>::storable_function(const T& object,
+                                                                        ReturnType (T::*method)(Args...) const) noexcept
+{
+    auto p = &object;
+    auto functor = [p, method](Args... args) -> ReturnType { return (*p.*method)(std::forward<Args>(args)...); };
+    storeFunctor(functor);
+}
+
+template <typename S, typename ReturnType, typename... Args>
+storable_function<S, signature<ReturnType, Args...>>::storable_function(const storable_function& other) noexcept
+    : m_vtable(other.m_vtable)
+{
+    m_vtable.copy(other, *this);
+}
+
+template <typename S, typename ReturnType, typename... Args>
+storable_function<S, signature<ReturnType, Args...>>::storable_function(storable_function&& other) noexcept
+    : m_vtable(other.m_vtable)
+{
+    m_vtable.move(other, *this);
+}
+
+template <typename S, typename ReturnType, typename... Args>
+storable_function<S, signature<ReturnType, Args...>>&
+storable_function<S, signature<ReturnType, Args...>>::operator=(const storable_function& rhs) noexcept
+{
+    if (&rhs != this)
+    {
+        // this vtable is needed for destroy, then changed to src vtable
+        m_vtable.destroy(*this);
+        m_function = nullptr; // only needed when the src has no object
+        m_vtable = rhs.m_vtable;
+        m_vtable.copy(rhs, *this);
+    }
+
+    return *this;
+}
+
+template <typename S, typename ReturnType, typename... Args>
+storable_function<S, signature<ReturnType, Args...>>&
+storable_function<S, signature<ReturnType, Args...>>::operator=(storable_function&& rhs) noexcept
+{
+    if (&rhs != this)
+    {
+        // this vtable is needed for destroy, then changed to source (rhs) vtable
+        m_vtable.destroy(*this);
+        m_function = nullptr; // only needed when the source has no object
+        m_vtable = rhs.m_vtable;
+        m_vtable.move(rhs, *this);
+    }
+
+    return *this;
+}
+
+template <typename S, typename ReturnType, typename... Args>
+storable_function<S, signature<ReturnType, Args...>>::~storable_function() noexcept
+{
+    m_vtable.destroy(*this);
+}
+
+
+template <typename S, typename ReturnType, typename... Args>
+ReturnType storable_function<S, signature<ReturnType, Args...>>::operator()(Args... args)
+{
+    auto r = m_function(std::forward<Args>(args)...);
```

Comment:
This is not perfect forwarding and here you create under some circumstances an additional move. A simple example would be:
```cpp
// copy but not movable
class B { 
   public:
    B() = default;
    B(const B &) = default;
    B(B &&) = delete;
    B &operator=(const B &) = default;
    B &operator=(B &&) = delete;
};

// function with argument as copy
void someFunction(B blubb) { ... }

int main() {
  storable_function<void(B)> func(someFunction);
  B bla;
  someFunction(bla); // works fine
  // does not compile since bla is not movable since std::forward<Args>(args)... enforces a move
  func(bla);
}
```

The solution is to implement perfect forwarding with variadic template arguments in the method and not in the class.
```cpp
template <typename S, typename ReturnType, typename... Args>
￼template <typename... FuncArgs>
ReturnType storable_function<S, signature<ReturnType, Args...>>::operator()(FuncArgs... args)
￼{
￼    auto r = m_function(std::forward<FuncArgs>(args)...);
}
```

But it would not be required to verify that `FuncArgs` and `Args` are the same since it would not compile if this wouldn't be the case.

Reply:
Need to think about the best approach here. Likely needs the additional variadic template parameter.

Reply:
The example you provided is wrongly parameterized `storable_function` is missing its storage type template parameter, but that is not the issue. Adding variadic arguments works and leads to perfect forwarding in `function`, but `function_ref` may introduce unwanted copies (in the function pointer case).

I am not entirely happy with the change for this reason, need to look closer at the function pointer case (maybe the ctor in `function_ref` can be avoided (might be not possible without C++17, see my comments regarding this).